server/background/process_qfx_upload.go

```diff
@@ -0,0 +1,467 @@
+package background
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/elliotcourant/gofx"
+	"github.com/getsentry/sentry-go"
+	"github.com/go-pg/pg/v10"
+	"github.com/monetr/monetr/server/crumbs"
+	"github.com/monetr/monetr/server/formats/qfx"
+	"github.com/monetr/monetr/server/internal/calc"
+	. "github.com/monetr/monetr/server/models"
+	"github.com/monetr/monetr/server/pubsub"
+	"github.com/monetr/monetr/server/repository"
+	"github.com/monetr/monetr/server/storage"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	ProcessQFXUpload = "ProcessQFXUpload"
+)
+
+var (
+	_ JobHandler        = &ProcessQFXUploadHandler{}
+	_ JobImplementation = &ProcessQFXUploadJob{}
+)
+
+type (
+	ProcessQFXUploadHandler struct {
+		log          *logrus.Entry
+		db           *pg.DB
+		publisher    pubsub.Publisher
+		files        storage.Storage
+		enqueuer     JobEnqueuer
+		unmarshaller JobUnmarshaller
+		clock        clock.Clock
+	}
+
+	ProcessQFXUploadArguments struct {
+		AccountId           ID[Account]           `json:"accountId"`
+		BankAccountId       ID[BankAccount]       `json:"bankAccountId"`
+		TransactionUploadId ID[TransactionUpload] `json:"transactionUploadId"`
+	}
+
+	ProcessQFXUploadJob struct {
+		args      ProcessQFXUploadArguments
+		log       *logrus.Entry
+		repo      repository.BaseRepository
+		files     storage.Storage
+		publisher pubsub.Publisher
+		enqueuer  JobEnqueuer
+		clock     clock.Clock
+		timezone  *time.Location
+
+		upload                *TransactionUpload
+		file                  *File
+		data                  *gofx.OFX
+		statementTransactions []gofx.StatementTransaction
+		existingTransactions  map[string]Transaction
+	}
+)
+
+func NewProcessQFXUploadHandler(
+	log *logrus.Entry,
+	db *pg.DB,
+	clock clock.Clock,
+	files storage.Storage,
+	publisher pubsub.Publisher,
+	enqueuer JobEnqueuer,
+) *ProcessQFXUploadHandler {
+	return &ProcessQFXUploadHandler{
+		log:          log,
+		db:           db,
+		publisher:    publisher,
+		files:        files,
+		enqueuer:     enqueuer,
+		unmarshaller: DefaultJobUnmarshaller,
+		clock:        clock,
+	}
+}
+
+func (h *ProcessQFXUploadHandler) updateStatus(
+	ctx context.Context,
+	args ProcessQFXUploadArguments,
+	status TransactionUploadStatus,
+	errorMessage *string,
+) error {
+	log := h.log.WithContext(ctx).WithFields(logrus.Fields{
+		"accountId":           args.AccountId,
+		"bankAccountId":       args.BankAccountId,
+		"transactionUploadId": args.TransactionUploadId,
+	})
+	channel := fmt.Sprintf("account:%s:transaction_upload:%s:progress", args.AccountId, args.TransactionUploadId)
+	payload := string(status)
+	if err := h.publisher.Notify(ctx, channel, payload); err != nil {
+		return errors.Wrap(err, "failed to send progress notification for job")
+	}
+	log.WithFields(logrus.Fields{
+		"channel": channel,
+		"payload": payload,
+	}).Trace("sent progress notification for file upload")
```

Comment:
Should move the notification to send after the row is updated. Otherwise there might be a race condition in the client where they receive a progress notification and read the upload record, only to get stale data.